// vitest/dist/suite.js
import { context } from './context';
export const defaultSuite = suite('');
export const test = (name, fn) => (context.currentSuite || defaultSuite).test(name, fn);
export function clearContext() {
    context.suites.length = 0;
    defaultSuite.clear();
}
export function suite(suiteName, factory) {
    const queue = [];
    const factoryQueue = [];
    const suite = {
        name: suiteName,
        test,
        collect,
        clear,
    };
    function test(name, fn) {
        const task = {
            suite,
            name,
            fn,
        };
        queue.push(task);
    }
    function clear() {
        queue.length = 0;
        factoryQueue.length = 0;
    }
    async function collect() {
        factoryQueue.length = 0;
        if (factory)
            await factory(test);
        return [...factoryQueue, ...queue];
    }
    context.currentSuite = suite;
    context.suites.push(suite);
    return suite;
}
// alias
export const describe = suite;
export const it = test;
